Objects/unionobject.c

```diff
@@ -137,102 +137,155 @@ union_richcompare(PyObject *a, PyObject *b, int op)
+    Py_ssize_t args_length = PyTuple_GET_SIZE(args);
+
+    PyObject *tuple = PyTuple_New(args_length + 1);
+
+    if (tuple == NULL) {
         return NULL;
     }
-    // Add unique elements to an array.
-    Py_ssize_t added_items = 0;
-    for (Py_ssize_t i = 0; i < arg_length; i++) {
-        int is_duplicate = 0;
-        PyObject* i_element = PyTuple_GET_ITEM(args, i);
-        for (Py_ssize_t j = 0; j < added_items; j++) {
-            PyObject* j_element = PyTuple_GET_ITEM(new_args, j);
-            int is_ga = _PyGenericAlias_Check(i_element) &&
-                        _PyGenericAlias_Check(j_element);
-            // RichCompare to also deduplicate GenericAlias types (slower)
-            is_duplicate = is_ga ? PyObject_RichCompareBool(i_element, j_element, Py_EQ)
-                : i_element == j_element;
-            // Should only happen if RichCompare fails
+
+    int is_duplicate = 0;
+
+    for (int i = 0; i < args_length; i++) {
+        PyObject *arg = PyTuple_GET_ITEM(args, i);
+
+        if (!is_duplicate) {
+            is_duplicate = is_same(arg, right);
+
             if (is_duplicate < 0) {
-                Py_DECREF(args);
-                Py_DECREF(new_args);
+                Py_DECREF(tuple);
                 return NULL;
             }
-            if (is_duplicate)
-                break;
         }
-        if (!is_duplicate) {
-            Py_INCREF(i_element);
-            PyTuple_SET_ITEM(new_args, added_items, i_element);
-            added_items++;
+
+        Py_INCREF(arg);
+        PyTuple_SET_ITEM(tuple, i, arg);
+    }
+
+    if (!is_duplicate) {
+        Py_INCREF(right);
+        PyTuple_SET_ITEM(tuple, args_length, right);
+    } else {
+        if (_PyTuple_Resize(&tuple, args_length) < 0) {
+            Py_DECREF(tuple);
+            return NULL;
         }
     }
-    Py_DECREF(args);
-    _PyTuple_Resize(&new_args, added_items);
-    return new_args;
+
+    return tuple;
 }
 
-static int
-is_unionable(PyObject *obj)
-{
-    return (obj == Py_None ||
-        PyType_Check(obj) ||
-        _PyGenericAlias_Check(obj) ||
-        _PyUnion_Check(obj));
+static PyObject*
+merge_obj_and_union(PyObject *left, PyObject *right) {
+    PyObject* args = ((unionobject *) right)->args;
+    Py_ssize_t args_length = PyTuple_GET_SIZE(args);
+
+    PyObject *tuple = PyTuple_New(args_length + 1);
+
+    if (tuple == NULL) {
+        return NULL;
+    }
+
+    Py_INCREF(left);
+    PyTuple_SET_ITEM(tuple, 0, left);
+
+    Py_ssize_t pos = 1;
+
+    for (int i = 0; i < args_length; i++) {
+        PyObject *arg = PyTuple_GET_ITEM(args, i);
+        int is_duplicate = is_same(left, arg);
+
+        if (is_duplicate < 0) {
+            Py_DECREF(tuple);
+            return NULL;
+        }
+        if (!is_duplicate) {
+            Py_INCREF(arg);
+            PyTuple_SET_ITEM(tuple, pos, arg);
+            pos++;
+        }
+    }
+
+    if(_PyTuple_Resize(&tuple, pos) < 0)  {
```

Comment:
PEP 7: one space between keywords like **if**, **for** and the following left paren

```suggestion
    if (_PyTuple_Resize(&tuple, pos) < 0)  {
```